lib/bap_types/bap_const_propagation.ml

```diff
@@ -0,0 +1,93 @@
+open Core_kernel.Std
+open Bap_common
+open Bap_bil
+open Bap_visitor
+open Stmt
+
+module Var = Bap_var
+module Word = Bitvector
+module Exp_helpers = Bap_helpers.Exp
+
+let is_assigned = Bap_helpers.is_assigned
+
+(** [remove_forward_defs bil defs]
+    for every [x := exp] in [bil], where [exp] depends on some
+    [y1 .. yN] redefined later then [x],
+    removes both [x] and [y1 .. yN] from defs *)
+let remove_forward_defs bil defs =
+  (object
+    inherit [exp Var.Map.t] bil_visitor
+    method! enter_move var e defs =
+      let redefined = Set.filter (Exp_helpers.free_vars e)
```

Comment:
It is likely that you are misunderstanding `free_vars`. For example, `free_vars (y := x + let x = z in x+ 1)` will yield `x,z`.

Also, why is a set of variables that are redefined by a move statement is a subset of a free variables of that statement s.t. they are not assigned after in any following statement? Either it is a wrong name or a wrong implementation.